lib/galaxy/tools/__init__.py

```diff
@@ -2510,42 +2510,64 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
         )
 
 
-class FilterFailedDatasetsTool(DatabaseOperationTool):
-    tool_type = 'filter_failed_datasets_collection'
-    require_dataset_ok = False
+class FilterDatasetsTool(DatabaseOperationTool):
+
+    def get_new_elements(self, history, elements_to_copy):
+        new_elements = odict()
+        for dce in elements_to_copy:
+            element_identifier = dce.element_identifier
+            copied_value = dce.element_object.copy()
+            if getattr(copied_value, "history_content_type", None) == "dataset":
+                history.add_dataset(copied_value, set_hid=False)
+            new_elements[element_identifier] = copied_value
+        return new_elements
 
     def produce_outputs(self, trans, out_data, output_collections, incoming, history, **kwds):
-        hdca = incoming["input"]
+        collection = incoming["input"]
 
-        assert hdca.collection.collection_type == "list" or hdca.collection.collection_type == 'list:paired'
+        if hasattr(collection, 'element_object'):
+            # A list
+            elements = collection.element_object.elements
+            collection_type = collection.element_object.collection_type
+        else:
+            # A pair
+            elements = collection.collection.elements
+            collection_type = collection.collection.collection_type
+        # We only process list or paired collections. Higher order collection will be mapped over
+        assert collection_type in ("list", "paired")
+
+        elements_to_copy = []
+        for element in elements:
+            if self.element_is_valid(element):
+                elements_to_copy.append(element)
+            elif collection_type == 'paired':
+                # One of the pairs is not OK, skip creating output for this pair
+                return
```

Comment:
Is it OK to just `return` here or should this return e.g. `self._outputs_dict()` as `DatabaseOperationTool.produce_outputs()` does?

Reply:
Good point, it actually isn't